doc/er_B.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "ER_B" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBer_B\fR \- Sample a random graph from the Erdos\-Renyi model B
.
.SH "SYNOPSIS"
\fBer_B\fR \fIN\fR \fIp\fR [\fIfileout\fR]
.
.SH "DESCRIPTION"
\fBer_B\fR samples a random graph from the Erdos\-Renyi model B, i\.e\. a graph with \fIN\fR nodes where each of the <N(N\-1)/2> edges is created independently with probability \fIp\fR\. The program dumps the edge list of the resulting graph on output\. If the optional \fIfileout\fR is provided, the output is written on a file with that name\.
.
.SH "PARAMETERS"
.
.TP
\fIN\fR
Number of nodes in the final graph\.
.
.TP
\fIp\fR
Edge probability\.
.
.TP
\fIfileout\fR
The (optional) name of the filename where the edge list of the graph will be saved\.
.
.SH "EXAMPLES"
The following command:
.
.IP "" 4
.
.nf

      $ er_B 1000 0\.006
.
.fi
.
.IP "" 0
.
.P
samples an undirected random network with \fIN=1000\fR nodes where each possible edge is created with probability \fIp=0\.006\fR\. The output of the command \fBer_B\fR will be the edge\-list of the resulting graph, where each (undirected) edge is reported only once\. In order to be useful, such edge\-list should be saved into a file\. The following command:
.
.IP "" 4
.
.nf

      $ er_B 1000 0\.006 > er_B_1000_0\.006\.net
.
.fi
.
.IP "" 0
.
.P
will save the resulting graph in the file er_B_1000_0\.006\.net\. Notice the usage of the symbol ">" to redirect the output of the program to a file\.
.
.SH "SEE ALSO"
er_A(1), ws(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
P\. Erdos, & A\. Rényi, "On Random Graphs I", Publ\. Math\. Debrecen, 6, 290 (1959)
.
.IP "\(bu" 4
P\.Erdos, & A\. Renyi, "On the evolution of random graphs" Publ\. Math\. Inst\. Hungary\. Acad\. Sci\., 5, 17\-61 (1960)
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 3, Cambridge University Press (2017)
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Appendix 10, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/er/er_B.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 *  This program samples a graph from the Erdos-Renyi random graph
 *  model B, i.e., the ensemble of graphs with N nodes where each of
 *  the possible N(N-1)/2 edges is created with probability equal to
 *  p, and independently from all the other edges.
 *
 *
 *  References: 
 *
 *  [1] Erdős, P. & Rényi, A. On Random Graphs I Publ. Math. Debrecen,
 *      6, 290 (1959)
 *
 *  [2] Erdos, P. & Renyi, A. On the evolution of random graphs
 *      Publ. Math. Inst. Hungary. Acad. Sci., 5, 17-61 (1960)
 *
 */



#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void usage(char *argv[]){
  printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-    er_B    -*-                      **\n"
         "**                                                                **\n"
         "**  sample an Erdos-Renyi graph of N nodes and edge probability p **\n"
         "**                                                                **\n"
         "**                  the output is an I,J matrix:                  **\n"
         "**                            I_1 J_1                             **\n"
         "**                            I_2 J_2                             **\n"
         "**                            I_3 J_3                             **\n"
         "**                            ... ...                             **\n"
         "**                            I_K J_K                             **\n"
         "**                                                                **\n"
         "**               (output is to stdout or to a file)               **\n"
         "**                                                                **\n"
         "**      If a filename is specified as fourth parameter then       **\n"
         "**         the degree sequence is dumped out to that file         **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " (c) Vincenzo Nicosia 2010-2017 (v.nicosia@qmul.ac.uk)\n\n"
         "********************************************************************\n\n"
         );
  printf("Usage: %s <N> <p> [<filout>]\n\n" , argv[0]);
}



int main(int argc, char *argv[]){

  unsigned int N, i, j;;
  FILE *fileout;
  double p, val;
  char error_str [256];

  if(argc < 3){
    usage(argv);
    exit(1);
  }
  
  N = atoi(argv[1]);
  p = atof(argv[2]);
  
  
  if (argc >3){
    fileout = fopen(argv[3], "w+");
    if (!fileout){
      sprintf(error_str, "Error opening file %s", argv[3]);
      perror(error_str);
      exit(2);
    }
  }
  else{
    fileout = stdout;
  }
  
  srand(time(NULL));

  /* Loop over all nodes...*/
  for(i=0; i<N; i++){
    /* Loop over the nodes with label larger than i...*/
    for(j=i+1; j<N; j++){
      val = rand() * 1.0 / RAND_MAX;
      /* if the sample is smaller than p, create the edge */
      if (val < p){
        fprintf(fileout, "%d %d\n", i, j);
      }
    }
  }
  fclose(fileout);
}
